Unable to update product images on Pro platform - product_id field _id missing org after barcode change

The producer coppola-industria-alimentare-srl have been trying to replace the existing image for a product on Open Food Facts, but the new image never saves, the upload just keeps **loading indefinitely.**

Here’s what they’ve tried so far:
- Uploading a .png file
- Switching to a lighter .jpg file

Neither option worked. Now, even attempting to revert to the original image is not possible, so the products currently have no images displayed. Their internet connection is stable, so I don’t believe it’s a connectivity issue.

Could you please investigate this problem? It seems like there might be a bug preventing image updates and restorations.

Example of a product were the issue occurs: https://world.pro.openfoodfacts.org/org/org-coppola-industria-alimentare-srl/product/5060323905388/organic-linguine-fontana-formiello

The issue is that the org_id is not passed to (or not inferred by) process_image_crop:

[2026/04/30 09:58:31] [48641] [ERROR] /srv/off-pro/lib/ProductOpener/Images.pm 1724 [cannot read image] {"request":"EVExRLFiFBGTaz9R","product_id":"5060323905388","imgid":1,"path":"/srv/off-pro/html/images/products/506/032/390/5388/1.jpg","error":"Exception 435: unable to open image `/srv/off-pro/html/images/products/506/032/390/5388/1.jpg': No such file or directory @ error/blob.c/OpenBlob/3029","code":"5060323905388","source_path":"/srv/off-pro/html/images/products/506/032/390/5388/1.jpg","id":"ingredients_en"}

Investigating.

--- a/lib/ProductOpener/Products.pm
+++ b/lib/ProductOpener/Products.pm
@@ -1151,7 +1151,9 @@ sub change_product_code ($product_ref, $new_code) {
 	}
 	else {
 		# check that the new code is available
-		if (object_exists("$BASE_DIRS{PRODUCTS}/" . product_path_from_id($new_code) . "/product")) {
+		# On the pro platform, product_id includes the owner prefix (e.g. org-xxx/code)
+		my $new_product_id = product_id_for_owner($Owner_id, $new_code);
+		if (object_exists("$BASE_DIRS{PRODUCTS}/" . product_path_from_id($new_product_id) . "/product")) {
 			$log->warn("cannot change product code, because the new code already exists",
 				{code => $code, new_code => $new_code})
 				if $log->is_warn();
@@ -1435,7 +1437,8 @@ sub store_product ($user_id, $product_ref, $comment, $client_id = undef) {
 					}
 				);
 			}
-			$product_ref->{_id} = $product_ref->{code} . '';    # treat id as string;
+			# On the pro platform, the _id includes the owner prefix (e.g. org-xxx/code)
+			$product_ref->{_id} = product_id_for_owner($Owner_id, $product_ref->{code}) . '';    # treat id as string;
 		}
 
 		if (object_path_exists("$BASE_DIRS{PRODUCTS}/$path")) {
@@ -1491,7 +1494,8 @@ sub store_product ($user_id, $product_ref, $comment, $client_id = undef) {
 				}
 			);
 
-			$product_ref->{_id} = $product_ref->{code} . '';    # treat id as string;
+			# On the pro platform, the _id includes the owner prefix (e.g. org-xxx/code)
+			$product_ref->{_id} = product_id_for_owner($Owner_id, $product_ref->{code}) . '';    # treat id as string;
 
 		}
 

--- a/lib/ProductOpener/Products.pm
+++ b/lib/ProductOpener/Products.pm
@@ -1152,7 +1152,10 @@ sub change_product_code ($product_ref, $new_code) {
 	else {
 		# check that the new code is available
 		# On the pro platform, product_id includes the owner prefix (e.g. org-xxx/code)
-		my $new_product_id = product_id_for_owner($Owner_id, $new_code);
+		# Use the product owner instead of the current Owner_id, which can differ for moderators
+		# or when editing products from another organization. Fallback to Owner_id if needed.
+		my $product_owner_id = $product_ref->{owner} // $Owner_id;
+		my $new_product_id = product_id_for_owner($product_owner_id, $new_code);
 		if (object_exists("$BASE_DIRS{PRODUCTS}/" . product_path_from_id($new_product_id) . "/product")) {
 			$log->warn("cannot change product code, because the new code already exists",
 				{code => $code, new_code => $new_code})

--- a/lib/ProductOpener/Products.pm
+++ b/lib/ProductOpener/Products.pm
@@ -1441,7 +1441,9 @@ sub store_product ($user_id, $product_ref, $comment, $client_id = undef) {
 				);
 			}
 			# On the pro platform, the _id includes the owner prefix (e.g. org-xxx/code)
-			$product_ref->{_id} = product_id_for_owner($Owner_id, $product_ref->{code}) . '';    # treat id as string;
+			my $product_owner_id = $product_ref->{owner} // $Owner_id;
+			$product_ref->{_id}
+				= product_id_for_owner($product_owner_id, $product_ref->{code}) . '';    # treat id as string;
 		}
 
 		if (object_path_exists("$BASE_DIRS{PRODUCTS}/$path")) {
@@ -1498,7 +1500,9 @@ sub store_product ($user_id, $product_ref, $comment, $client_id = undef) {
 			);
 
 			# On the pro platform, the _id includes the owner prefix (e.g. org-xxx/code)
-			$product_ref->{_id} = product_id_for_owner($Owner_id, $product_ref->{code}) . '';    # treat id as string;
+			my $product_owner_id = $product_ref->{owner} // $Owner_id;
+			$product_ref->{_id}
+				= product_id_for_owner($product_owner_id, $product_ref->{code}) . '';    # treat id as string;
 
 		}
 

--- a/scripts/update_all_products.pl
+++ b/scripts/update_all_products.pl
@@ -406,7 +406,7 @@ my $products_collection = get_products_collection({obsolete => $obsolete, timeou
 # Collections for saving current / obsolete products
 my %products_collections = (
 	current => get_products_collection({timeout => $socket_timeout_ms}),
-	obsolete => get_products_collection({obsolete => $obsolete, timeout => $socket_timeout_ms}),
+	obsolete => get_products_collection({obsolete => 1, timeout => $socket_timeout_ms}),
 );
 
 my $products_count = "";

--- a/scripts/update_all_products.pl
+++ b/scripts/update_all_products.pl
@@ -49,6 +49,7 @@ it is likely that the MongoDB cursor of products to be updated will expire, and 
 --check-quality	run quality checks
 --compute-codes
 --fix-serving-size-mg-to-ml
+--fix-product-id-missing-owner-prefix	fix products on the pro platform where the _id field is missing the owner prefix (e.g. code instead of org-xxx/code)
 --index		specifies that the keywords used by the free text search function (name, brand etc.) need to be reindexed. -- TBD
 --user		create a separate .sto file and log the change in the product history, with the corresponding user
 --team		optional team for the user that is credited with the change
@@ -150,6 +151,7 @@ my $compute_main_countries = '';
 my $prefix_packaging_tags_with_language = '';
 my $fix_non_string_ids = '';
 my $fix_non_string_codes = '';
+my $fix_product_id_missing_owner_prefix = '';
 my $fix_string_last_modified_t = '';
 my $assign_ciqual_codes = '';
 my $obsolete = 0;
@@ -191,6 +193,7 @@ GetOptions(
 	"fix-rev-not-incremented" => \$fix_rev_not_incremented,
 	"fix-non-string-ids" => \$fix_non_string_ids,
 	"fix-non-string-codes" => \$fix_non_string_codes,
+	"fix-product-id-missing-owner-prefix" => \$fix_product_id_missing_owner_prefix,
 	"fix-string-last-modified-t" => \$fix_string_last_modified_t,
 	"force-new-version" => \$force_new_version,
 	"user-id=s" => \$User_id,
@@ -279,6 +282,7 @@ if (    (not $process_ingredients)
 	and (not $fix_non_string_ids)
 	and (not $fix_non_string_codes)
 	and (not $fix_string_last_modified_t)
+	and (not $fix_product_id_missing_owner_prefix)
 	and (not $compute_sort_key)
 	and (not $remove_team)
 	and (not $remove_category)
@@ -326,6 +330,16 @@ my $query_ref = {};
 
 add_params_to_query($query_params_ref, $query_ref);
 
+# Query products on the pro platform where _id is missing the owner prefix
+# (e.g. _id = "5060323905388" instead of "org-xxx/5060323905388")
+# This can happen after a barcode change due to a bug in store_product()
+if ($fix_product_id_missing_owner_prefix) {
+	# On the pro platform, valid _id values always contain '/' (owner/code format)
+	# Barcodes (all digits) do not contain '/', so we look for _id starting with a digit
+	$query_ref->{_id} = {'$regex' => '^\d'};
+	$query_ref->{owner} = {'$exists' => 1};
+}
+
 # Query products that have the _id field stored as a number
 if ($fix_non_string_ids) {
 	$query_ref->{_id} = {'$type' => "long"};
@@ -349,7 +363,10 @@ if ($add_product_type) {
 # On the producers platform, require --query owners_tags to be set, or the --all-owners field to be set.
 
 if ((defined $server_options{private_products}) and ($server_options{private_products})) {
-	if ((not $all_owners) and (not defined $query_ref->{owners_tags})) {
+	if (    (not $all_owners)
+		and (not defined $query_ref->{owners_tags})
+		and (not $fix_product_id_missing_owner_prefix))
+	{
 		print STDERR "On producers platform, --query owners_tags=... or --all-owners must be set.\n";
 		exit();
 	}
@@ -468,6 +485,43 @@ while (my $product_ref = $cursor->next) {
 	}
 
 	next if $just_print_codes;
+
+	# Fix products on the pro platform where _id is missing the owner prefix
+	if ($fix_product_id_missing_owner_prefix) {
+		my $owner = $product_ref->{owner};
+		if ((not defined $owner) or ($owner eq '')) {
+			print STDERR ". Error: no owner field for product $code with _id $productid\n";
+			next;
+		}
+		my $correct_product_id = $owner . "/" . $code;
+		print STDERR " - fixing _id: $productid -> $correct_product_id";
+		if (!$pretend) {
+			# Read from disk using the correct owner-prefixed product_id
+			my $fixed_product_ref = retrieve_product($correct_product_id);
+			if (defined $fixed_product_ref) {
+				# Fix the _id in the product data
+				$fixed_product_ref->{_id} = $correct_product_id;
+				my $fixed_path = product_path($fixed_product_ref);
+				# Update the .sto file with the corrected _id
+				store_object("$BASE_DIRS{PRODUCTS}/$fixed_path/product", $fixed_product_ref);
+				# Delete the old MongoDB document with the wrong _id
+				$products_collection->delete_one({"_id" => $productid});
+				# Insert the product with the correct _id
+				my $collection = "current";
+				if ($fixed_product_ref->{obsolete}) {
+					$collection = "obsolete";
+				}
+				$products_collections{$collection}
+					->replace_one({"_id" => $correct_product_id}, $fixed_product_ref, {upsert => 1});
+				$products_changed++;
+				print STDERR ". Fixed";
+			}
+			else {
+				print STDERR ". Error: could not load product from disk using product_id $correct_product_id";
+			}
+		}
+		next;
+	}
 
 	if (!$mongodb_to_mongodb) {
 		# read product data from .sto file

